pkg/grpc/api.go

```diff
@@ -691,6 +702,105 @@ func (a *api) getStateStore(name string) (state.Store, error) {
+			component = statestore
+		}
+	case "pubsub":
+		if pubsub := a.pubsubAdapter.GetPubSub(componentName); pubsub != nil {
+			component = pubsub
+		}
+	case "secretstores":
+		if secretstore := a.secretStores[componentName]; secretstore != nil {
+			component = secretstore
+		}
+	case "bindings":
+		if inputbinding := a.inputBindings[componentName]; inputbinding != nil {
+			component = inputbinding
+		} else if outputbinding := a.outputBindings[componentName]; outputbinding != nil {
+			component = outputbinding
+		}
+	}
+	return
+}
+
+func (a *api) GetAllComponentsHealthAlpha1(ctx context.Context, in *runtimev1pb.AllComponentsHealthRequest) (*runtimev1pb.AllComponentsHealthResponse, error) {
+	i := 0
+	components := a.getComponentsFn()
+	hresp := &runtimev1pb.AllComponentsHealthResponse{
+		Results: make([]*runtimev1pb.ComponentHealthResponseItem, len(components)),
+	}
+	for _, comp := range components {
+		a.allComponentsHealthResponsePopulator(strings.Split(comp.Spec.Type, ".")[0], hresp, i, comp.Name)
+		i++
+	}
+	return hresp, nil
+}
+
+func (a *api) allComponentsHealthResponsePopulator(componentType string, hresp *runtimev1pb.AllComponentsHealthResponse, ind int, name string) {
+	status, errStr, message, _ := a.CheckHealthUtil(componentType, name)
+
+	hresp.Results[ind] = &runtimev1pb.ComponentHealthResponseItem{
+		ComponentName: name,
+		Type:          componentType,
+		Status:        status,
+		ErrorCode:     errStr,
+		Message:       message,
+	}
+}
+
+func (a *api) GetComponentHealthAlpha1(ctx context.Context, in *runtimev1pb.ComponentHealthRequest) (*runtimev1pb.ComponentHealthResponse, error) {
+	componentName := in.ComponentName
+	components := a.getComponentsFn()
+	for _, comp := range components {
+		if componentName == comp.Name {
+			status, errStr, message, err := a.CheckHealthUtil(strings.Split(comp.Spec.Type, ".")[0], comp.Name)
+			return &runtimev1pb.ComponentHealthResponse{
+				Status:    status,
+				ErrorCode: errStr,
+				Message:   message,
+			}, err
+		}
+	}
+	err := status.Errorf(codes.InvalidArgument, messages.ErrComponentNotFound)
+	apiServerLogger.Debug(err)
+
+	return &runtimev1pb.ComponentHealthResponse{
+		Status:    utils.StatusUndefined,
+		ErrorCode: messages.ErrComponentNotFound,
+	}, err
+}
+
+func (a *api) CheckHealthUtil(componentKind string, componentName string) (
+	healthStatus string, errorCode string, message string, err error,
+) {
+	component := a.getComponent(componentKind, componentName)
+
+	if component == nil {
+		err := status.Errorf(codes.InvalidArgument, messages.ErrComponentNotFound)
+		apiServerLogger.Debug(err)
+
+		return utils.StatusUndefined, messages.ErrComponentNotFound, "", err
+	}
+
+	if pinger, ok := component.(health.Pinger); ok {
+		pingErr := pinger.Ping()
+		if pingErr != nil {
+			err := status.Errorf(codes.Unknown, messages.ErrHealthNotOk)
+			apiServerLogger.Debug(pingErr)
+
+			return utils.StatusNotOk, messages.ErrHealthNotOk, pingErr.Error(), err
+		}
```

Comment:
return OK after this line inside the if block for pinged interface check ? Then the else nesting is not needed?

Reply:
Done